packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js

```diff
@@ -0,0 +1,329 @@
+        );
+      }
+
+      const container = document.createElement('div');
+      // attach to body so events works
+      document.body.appendChild(container);
+      let calledCounter = 0;
+      act(() => {
+        ReactDOM.render(
+          <App
+            callback={val => {
+              calledCounter = val;
+            }}
+          />,
+          container,
+        );
+      });
+      const button = document.getElementById('button');
+      function click() {
+        button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
+      }
+
+      act(() => {
+        click();
+        click();
+        click();
+      });
+      expect(calledCounter).toBe(3);
+      act(click);
+      expect(calledCounter).toBe(4);
+      act(click);
+      expect(calledCounter).toBe(5);
+      expect(button.innerHTML).toBe('5');
+
+      document.body.removeChild(container);
+    });
+
+    it('should flush effects recursively', () => {
+      function App() {
+        let [ctr, setCtr] = React.useState(0);
+        React.useEffect(() => {
+          if (ctr < 5) {
+            setCtr(x => x + 1);
+          }
+        });
+        return ctr;
+      }
+
+      const container = document.createElement('div');
+      act(() => {
+        ReactDOM.render(<App />, container);
+      });
+
+      expect(container.innerHTML).toBe('5');
+    });
+
+    it('detects setState being called outside of act(...)', () => {
+      let setValue = null;
+      function App() {
+        let [value, _setValue] = React.useState(0);
+        setValue = _setValue;
+        return (
+          <button id="button" onClick={() => setValue(2)}>
+            {value}
+          </button>
+        );
+      }
+      const container = document.createElement('div');
+      document.body.appendChild(container);
+      let button;
+      act(() => {
+        ReactDOM.render(<App />, container);
+        button = container.querySelector('#button');
+        button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
+      });
+      expect(button.innerHTML).toBe('2');
+      expect(() => setValue(1)).toWarnDev([
+        'An update to App inside a test was not wrapped in act(...).',
+      ]);
+      document.body.removeChild(container);
+    });
+    describe('fake timers', () => {
+      beforeEach(() => {
+        jest.useFakeTimers();
+      });
+      afterEach(() => {
+        jest.useRealTimers();
+      });
+      it('lets a ticker update', () => {
+        function App() {
+          let [toggle, setToggle] = React.useState(0);
+          React.useEffect(() => {
+            let timeout = setTimeout(() => {
+              setToggle(1);
+            }, 200);
+            return () => clearTimeout(timeout);
+          }, []);
+          return toggle;
+        }
+        const container = document.createElement('div');
+
+        act(() => {
+          ReactDOM.render(<App />, container);
+        });
+        act(() => {
+          jest.runAllTimers();
```

Comment:
If you replace the `setTimeout` in the effect with `sleep(200).then(() => setToggle(1))` then the test will fail, because the update happens in a microtask. That's why it's important that you always await `act`. We need another test case that shows that.

Reply:
I'm not sure that's true with fakeTimers, but I'll verify.

Reply:
Even if it's not true for fake timers (though I think it is) it would be true for other types of events, like network.

Reply:
>most people who use timers are going to use the fake timers, anyway.

I don't believe this to be the case in open source. I've mostly seen people being wary of mocking timers for different reasons (often emotional ones like "doesn't feel right") and we need to meet them where they are.

Reply:
so I tried, and - 
- in our setup (and likely oss too), this will trigger the warning. the fix for this is simple - use the async version to catch microtasks like so 
```jsx
await act(async () => {
  jest.runAllTimers()
})
```
- I'd assumed we had a setup similar to fb's, where the changed code *will not* trigger the warning (I just tested)

I'm updating the test to show this.

Reply:
this is now better since we wait for a whole task.